subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c

```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+			continue;
+		}
+
+		for (size_t j = 0; j < account_key_count; j++) {
+			if (fp_bonds[i].nv.account_key_metadata_id ==
+			    ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[j], ID)) {
+				account_key_found = true;
+				break;
+			}
+		}
+
+		if (!account_key_found) {
+			LOG_DBG("Removing bond associated with invalid Account Key");
+			err = bond_remove_completely(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
+	}
+}
+
+static void bonds_conn_ctx_reset(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		fp_bonds[i].conn_ctx = NULL;
+	}
+}
+
+static void bt_bonds_find(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		fp_bonds[i].bonded = bt_request_cb->is_addr_bonded(&fp_bonds[i].nv.addr);
+		if (fp_bonds[i].bonded) {
+			LOG_DBG("Loaded bond has been found in the Bluetooth bond list");
+		}
+	}
+}
+
+static void bond_duplicates_handle(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		bool duplicate_found = false;
+		int err;
+
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY)) {
+			continue;
+		}
+
+		for (size_t j = i + 1; j < ARRAY_SIZE(fp_bonds); j++) {
+			if (bt_addr_le_eq(&fp_bonds[j].nv.addr, &fp_bonds[i].nv.addr)) {
+				/* Error during Fast Pair Procedure. */
+				duplicate_found = true;
+				LOG_DBG("Removing both duplicated bond entries at bootup");
+				err = fp_bond_delete(&fp_bonds[j]);
+				if (err) {
+					LOG_ERR("Failed to delete bond (err %d)", err);
+				}
+			}
+		}
+
+		if (duplicate_found) {
+			err = bond_remove_completely(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
+	}
+}
+
+static void invalid_bonds_purge(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		int err;
+
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY)) {
+			continue;
+		}
+
+		if (!fp_bonds[i].bonded) {
+			LOG_DBG("Deleting not bonded entry");
+			err = fp_bond_delete(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to delete bond (err %d)", err);
+			}
+
+			continue;
+		}
+
+		if (fp_bonds[i].nv.account_key_metadata_id == BOND_AK_ID_UNKNOWN) {
+			/* Fast Pair Procedure unfinished. */
+			err = bond_remove_completely(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
+	}
+}
+
+static void validate_bonds(void)
+{
+	/* Assert that zero-initialized addresses are equal to BT_ADDR_LE_ANY. */
+	static const bt_addr_le_t addr;
+
+	__ASSERT_NO_MSG(bt_addr_le_eq(&addr, BT_ADDR_LE_ANY));
+	ARG_UNUSED(addr);
+
+	/* If Fast Pair was disabled before the peer that performed the unsuccessful Fast Pair
+	 * Procedure disconnected, there may be a conn_ctx not set to NULL. All conn_ctx should be
+	 * set to NULL at this point.
+	 */
+	bonds_conn_ctx_reset();
+
+	bt_bonds_find();
+
+	bond_duplicates_handle();
+
+	invalid_bonds_purge();
+
+	bonds_with_invalid_id_remove_completely();
+}
+
 static int fp_settings_validate(void)
```

Comment:
Consider making functions like `fp_settings_validate_ak` and `fp_settings_validate_fp_bond` (to maintain the same level of abstraction)